File: packages/pip/src/goodvibes_cli/steps/global_setup.py
"""Global (user-level) Claude Code setup (mirror of npm steps/global-setup.ts)."""
from __future__ import annotations

import hashlib
import json
import os
import pathlib
import re
import shutil
import subprocess
import sys

from goodvibes_cli.steps.copy_templates import list_template_files
from goodvibes_cli.steps.write_manifest import MANIFEST_PATH, USER_OWNED, USER_REMOVED, read_manifest
from goodvibes_cli.utils.json_merge import merge_managed_json, present_ids, shape_error, write_json
from goodvibes_cli.utils.proc import run
from goodvibes_cli.utils.scope import goodvibes_block
from goodvibes_cli.utils.safe_path import remove_retired

CONTEXT7_URL = "https://mcp.context7.com/mcp"


def claude_config_dir() -> pathlib.Path:
    return pathlib.Path(os.environ.get("CLAUDE_CONFIG_DIR") or pathlib.Path.home() / ".claude")


def _sha(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def register_context7(dry_run: bool) -> dict[str, str]:
    try:
        listed = subprocess.run(["claude", "mcp", "list"], capture_output=True, text=True, timeout=10)
        if re.search(r"^context7\b", listed.stdout, re.M):
            return {"status": "already-registered"}
        if dry_run:
            return {"status": "skipped", "reason": "dry run"}
        subprocess.run(
            ["claude", "mcp", "add", "--transport", "http", "--scope", "user", "context7", CONTEXT7_URL],
            capture_output=True, text=True, timeout=10, check=True,
        )
        return {"status": "registered"}
    except FileNotFoundError:
        return {"status": "skipped", "reason": f"claude CLI not found; run: claude mcp add --transport http --scope user context7 {CONTEXT7_URL}"}
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
        return {"status": "failed", "reason": str(e).splitlines()[0]}


def ensure_global_cli(version: str, dry_run: bool) -> dict[str, str]:
    """A plain `pip install` already puts goodvibes on PATH; `uvx`/`pipx run` do not, so install it as a uv tool."""
    found = shutil.which("goodvibes")
    # A goodvibes inside the active virtualenv is on PATH only while that venv is active.
    in_venv = found and sys.prefix != sys.base_prefix and pathlib.Path(found).resolve().is_relative_to(pathlib.Path(sys.prefix).resolve())
    if found and not in_venv:
        return {"status": "already-installed"}
    # >= not ==: uv stores the requirement, and a pin makes every later `uv tool upgrade` a no-op.
    manual = f'Install manually: uv tool install "goodvibes-cli>={version}"'
    if dry_run:
        return {"status": "skipped", "reason": f'dry run; would run uv tool install "goodvibes-cli>={version}"'}
    try:
        run(["uv", "tool", "install", f"goodvibes-cli>={version}"], capture_output=True, text=True, timeout=120, check=True)
        if shutil.which("goodvibes") is None:
            return {"status": "installed", "reason": "goodvibes is not on your PATH yet: run uv tool update-shell, then open a new terminal"}
        return {"status": "installed"}
    except FileNotFoundError:
        return {"status": "failed", "reason": f"uv not found. {manual} (or pip install goodvibes-cli)"}
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
        return {"status": "failed", "reason": f"{str(e).splitlines()[0]}. {manual}"}


def apply_global_config(template_dir: pathlib.Path, version: str, dry_run: bool, restore: bool = True) -> dict:
    """Write goodvibes-owned files into the Claude Code user config; a file the user edited since is kept.

    restore=False (update) leaves a recorded file the user deleted deleted; init passes True to bring it back.
    """
    cfg = claude_config_dir()
    prev = read_manifest(cfg) or {}
    prev_files = prev.get("files") or {}
    owned = [("rules/goodvibes.md", goodvibes_block((template_dir / "CLAUDE.md").read_text(encoding="utf-8")))]
    for rel in list_template_files(template_dir):
        rel = rel.replace("\\", "/")
        if rel.startswith(".claude/skills/"):
            owned.append((rel[len(".claude/"):], (template_dir / rel).read_text(encoding="utf-8")))

    result: dict = {"config_dir": str(cfg), "written": [], "kept": [], "removed": [], "retired": [], "settings_changes": [], "settings_error": None}
    files: dict[str, str] = {}
    for rel, content in owned:
        dest = cfg / rel
        recorded = prev_files.get(rel)
        if recorded == USER_REMOVED and dest.exists():
            # The user recreated it: theirs now, never overwritten (not even by init).
            result["kept"].append(rel)
            files[rel] = USER_OWNED
            continue
        if not restore and recorded == USER_REMOVED:
            files[rel] = USER_REMOVED
            continue
        if not restore and recorded and not dest.exists():
            result["removed"].append(rel)
            files[rel] = USER_REMOVED
            continue
        current = _sha(dest.read_text(encoding="utf-8")) if dest.exists() else None
        if current == _sha(content):
            files[rel] = current
            continue
        if current and current != recorded:
            result["kept"].append(rel)
            if recorded:
                files[rel] = recorded
            continue
        result["written"].append(rel)
        files[rel] = _sha(content)
        if not dry_run:
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_text(content, encoding="utf-8")

    # Skills goodvibes no longer ships: delete the copy it wrote; an edited copy is the user's and stays (untracked).
    owned_paths = {rel for rel, _ in owned}
    for rel, recorded in prev_files.items():
        if rel in owned_paths or not rel.startswith("skills/") or recorded in (USER_OWNED, USER_REMOVED):
            continue
        dest = cfg / rel
        if not dest.exists() or _sha(dest.read_text(encoding="utf-8")) != recorded:
            continue
        result["retired"].append(rel)
        if not dry_run:
            remove_retired(cfg, rel, "skills")

    tpl = json.loads((template_dir / ".claude" / "settings.json").read_text(encoding="utf-8"))
    settings_path = cfg / "settings.json"
    managed = dict(prev.get("managed") or {})
    try:
        user = json.loads(settings_path.read_text(encoding="utf-8")) if settings_path.exists() else {}
    except ValueError as e:
        user = None
        result["settings_error"] = f"{settings_path}: not valid JSON ({e}); left unchanged, fix it and re-run"
    shape = shape_error(".claude/settings.json", user) if isinstance(user, dict) else None
    if user is not None and not isinstance(user, dict):
        result["settings_error"] = f"{settings_path}: not a JSON object; left unchanged, fix it and re-run"
    elif shape:
        result["settings_error"] = f"{settings_path}: {shape}; left unchanged, fix it and re-run"
    elif user is not None:
        merged, changes = merge_managed_json(".claude/settings.json", tpl, user, managed.get("settings.json"))
        result["settings_changes"] = changes
        if not dry_run and changes:
            cfg.mkdir(parents=True, exist_ok=True)
            write_json(settings_path, merged)
        managed["settings.json"] = list(dict.fromkeys([*managed.get("settings.json", []), *present_ids(".claude/settings.json", tpl, merged)]))

    if not dry_run:
        cfg.mkdir(parents=True, exist_ok=True)
        write_json(cfg / MANIFEST_PATH, {"version": version, "scope": "global", "files": files, "managed": managed})
    return result


def format_global(g: dict, cli: dict | None, c7: dict | None) -> str:
    lines = [f"written: {f}" for f in g["written"]]
    lines += [f"kept (you edited it): {f}" for f in g["kept"]]
    lines += [f"{f}: removed by you, not re-added (run goodvibes init to restore)" for f in g.get("removed", [])]
    lines += [f"{f}: removed, no longer shipped by goodvibes" for f in g.get("retired", [])]
    lines += [f"settings.json {c}" for c in g["settings_changes"]]
    if g.get("settings_error"):
        lines.append(f"settings.json not changed: {g['settings_error']}")
    if c7:
        lines.append(f"context7 MCP: {c7['status']}" + (f" ({c7['reason']})" if c7.get("reason") else ""))
    if cli:
        lines.append(f"goodvibes CLI: {cli['status']}" + (f" ({cli['reason']})" if cli.get("reason") else ""))
    return "\n".join(lines) or "already up to date"

File: packages/npm/src/steps/global-setup.ts
import { readFile, writeFile, mkdir } from 'node:fs/promises'
import { existsSync } from 'node:fs'
import { createHash } from 'node:crypto'
import { homedir } from 'node:os'
import { dirname, join } from 'node:path'
import { execa } from 'execa'
import { listTemplateFiles } from './copy-templates.js'
import { readManifest, type Manifest, MANIFEST_PATH } from './write-manifest.js'
import { mergeManagedJson, presentIds, isJsonObject } from '../utils/json-merge.js'
import { writeFileAtomic } from '../utils/fs-safe.js'
import { goodvibesBlock } from '../utils/scope.js'
import { versionGte } from '../utils/sentinel-merge.js'

const CONTEXT7_URL = 'https://mcp.context7.com/mcp'

export function claudeConfigDir(): string {
  return process.env.CLAUDE_CONFIG_DIR || join(homedir(), '.claude')
}

const sha = (s: string) => createHash('sha256').update(s, 'utf8').digest('hex')

export type McpStatus = { status: 'registered' | 'already-registered' | 'skipped' | 'failed'; reason?: string }

export async function registerContext7(dryRun: boolean): Promise<McpStatus> {
  try {
    const { stdout } = await execa('claude', ['mcp', 'list'], { timeout: 10_000 })
    if (/^context7\b/m.test(stdout)) return { status: 'already-registered' }
    if (dryRun) return { status: 'skipped', reason: 'dry run' }
    await execa('claude', ['mcp', 'add', '--transport', 'http', '--scope', 'user', 'context7', CONTEXT7_URL], { timeout: 10_000 })
    return { status: 'registered' }
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
      return { status: 'skipped', reason: `claude CLI not found; run: claude mcp add --transport http --scope user context7 ${CONTEXT7_URL}` }
    }
    return { status: 'failed', reason: (e as Error).message.split('\n')[0] }
  }
}

export type CliStatus = { status: 'installed' | 'already-installed' | 'failed' | 'skipped'; reason?: string }

export async function ensureGlobalCli(version: string, dryRun: boolean): Promise<CliStatus> {
  try {
    const { stdout } = await execa('npm', ['ls', '-g', 'goodvibes-cli', '--depth=0', '--json'], { reject: false, timeout: 30_000 })
    const current = JSON.parse(stdout || '{}').dependencies?.['goodvibes-cli']?.version
    // Never downgrade: an older npx run must not replace a newer global install.
    if (current && versionGte(current, version)) return { status: 'already-installed' }
    if (dryRun) return { status: 'skipped', reason: `dry run; would run npm install -g goodvibes-cli@${version}` }
    await execa('npm', ['install', '-g', `goodvibes-cli@${version}`], { timeout: 120_000 })
    return { status: 'installed' }
  } catch (e) {
    const msg = (e as Error).message
    const hint = /EACCES|permission/i.test(msg)
      ? 'npm cannot write its global folder; see https://docs.npmjs.com/resolving-eacces-permissions-errors-when-installing-packages-globally'
      : msg.split('\n')[0]
    return { status: 'failed', reason: `${hint}. Install manually: npm install -g goodvibes-cli@${version}` }
  }
}

export type GlobalResult = {
  configDir: string
  written: string[]
  kept: string[]
  removed: string[]
  settingsChanges: string[]
  settingsError?: string
}

// Writes goodvibes-owned files into the Claude Code user config; a file the user edited since goodvibes wrote it is kept.
export async function applyGlobalConfig(templateDir: string, version: string, dryRun: boolean): Promise<GlobalResult> {
  const cfg = claudeConfigDir()
  const prev: Manifest | null = await readManifest(cfg)
  const owned: [string, string][] = [['rules/goodvibes.md', goodvibesBlock(await readFile(join(templateDir, 'CLAUDE.md'), 'utf-8'))]]
  for (const rel of (await listTemplateFiles(templateDir)).map(f => f.split('\\').join('/'))) {
    if (rel.startsWith('.claude/skills/')) owned.push([rel.slice('.claude/'.length), await readFile(join(templateDir, rel), 'utf-8')])
  }

  const result: GlobalResult = { configDir: cfg, written: [], kept: [], removed: [], settingsChanges: [] }
  const files: Record<string, string> = {}
  for (const [rel, content] of owned) {
    const dest = join(cfg, rel)
    const recorded = prev?.files[rel]
    // Tracked but gone: the user deleted it, so it is neither rewritten nor kept in the manifest.
    if (recorded && !existsSync(dest)) {
      result.removed.push(rel)
      continue
    }
    if (existsSync(dest) && sha(await readFile(dest, 'utf-8')) !== recorded) {
      result.kept.push(rel)
      if (recorded) files[rel] = recorded
      continue
    }
    result.written.push(rel)
    files[rel] = sha(content)
    if (!dryRun) {
      await mkdir(dirname(dest), { recursive: true })
      await writeFile(dest, content, 'utf-8')
    }
  }

  const tpl = JSON.parse(await readFile(join(templateDir, '.claude', 'settings.json'), 'utf-8'))
  const settingsPath = join(cfg, 'settings.json')
  let managed = prev?.managed ?? {}
  let user: unknown
  try {
    user = existsSync(settingsPath) ? JSON.parse(await readFile(settingsPath, 'utf-8')) : {}
    if (!isJsonObject(user)) result.settingsError = `${settingsPath}: not a JSON object; left unchanged, fix it and re-run`
  } catch (e) {
    result.settingsError = `${settingsPath}: not valid JSON (${(e as Error).message}); left unchanged, fix it and re-run`
  }
  if (isJsonObject(user)) {
    const { merged, changes } = mergeManagedJson('.claude/settings.json', tpl, user, managed['settings.json'])
    result.settingsChanges = changes
    if (!dryRun && changes.length > 0) {
      await mkdir(cfg, { recursive: true })
      await writeFileAtomic(settingsPath, JSON.stringify(merged, null, 2) + '\n')
    }
    managed = { ...managed, 'settings.json': [...new Set([...(managed['settings.json'] ?? []), ...presentIds('.claude/settings.json', tpl, merged)])] }
  }

  if (!dryRun) {
    await mkdir(cfg, { recursive: true })
    await writeFileAtomic(join(cfg, MANIFEST_PATH), JSON.stringify({ version, scope: 'global', files, managed }, null, 2) + '\n')
  }
  return result
}

export function formatGlobal(g: GlobalResult, cli: CliStatus | undefined, c7: McpStatus | undefined): string {
  const lines = [
    ...g.written.map(f => `written: ${f}`),
    ...g.kept.map(f => `kept (you edited it): ${f}`),
    ...g.removed.map(f => `${f}: removed by you, not re-added (run goodvibes init to restore)`),
    ...g.settingsChanges.map(c => `settings.json ${c}`),
    ...(g.settingsError ? [`settings.json not changed: ${g.settingsError}`] : []),
  ]
  if (c7) lines.push(`context7 MCP: ${c7.status}${c7.reason ? ` (${c7.reason})` : ''}`)
  if (cli) lines.push(`goodvibes CLI: ${cli.status}${cli.reason ? ` (${cli.reason})` : ''}`)
  return lines.join('\n') || 'already up to date'
}
